Rework map generation window

AaronVanGeffen opened the pull request:
This PR reworks the map generation window, as discussed in #22719. After this PR, we can gradually add new features, some of which may be inspired by, or even ported from, OpenLoco's landscape generator.

![Unnamed park 2024-09-11 17-21-12](https://github.com/user-attachments/assets/9ae116e9-6172-4cb6-b13d-6e9f377a2781) ![Unnamed park 2024-09-23 14-49-30](https://github.com/user-attachments/assets/ebd8e0bd-4f39-4a2f-947e-47ae8aecc0ba) ![Unnamed park 2024-09-11 17-21-19](https://github.com/user-attachments/assets/50e29e6e-c587-4ffa-b895-15c72abf1c16) ![Unnamed park 2024-09-11 17-21-21](https://github.com/user-attachments/assets/44b2c333-b220-4480-a9ea-a291540b8db1) ![Unnamed park 2024-09-11 17-21-22](https://github.com/user-attachments/assets/e59f0122-e6d5-4c30-aebe-dafa1ef20d25)

To do:
- [x] Introduce a base tab for the map size
- [x] Add generator selection to the initial tab
- [x] Consolidate generator buttons in shared widget definitions.
- [x] Add a terrain tab for the current terrain options, and only show them there.
- [x] Add tree placement as an option to all generators, not just simplex.
- [x] Add settings for number of trees, min/max tree altitude.
- [x] Add a water tab for the current water options, and only show them there.
- [x] Add beach texturing as an option to all generators, not just simplex.
- [x] Add tile edge smoothing setting as an option to all generators, not just the height map file one.
- [x] Unify min/max land height settings in terrain tab, instead having one per generator.
- [x] Widen the window, reduce height, space out widgets.
- [x] Change window captions based on the selected tab.
- [x] Move settings for the individual generators to the base tab.
- [x] Remove the tabs for the individual generators.
- [x] Disable 'generate' button for height map generator when no file has been loaded yet.
- [x] Display heights in metres/feet/units instead of the current mishmash
- [x] Add text input for land and tree altitude limits
- [x] Change default height values/limits to something more sensible
- [x] Consolidate base height setting in min land height setting
- [x] Investigate min/max land height limits for simplex generator -- needs div 2?
- [x] Add 'randomise' options to simplex options instead of second 'random' simplex generator? Or just better defaults without the extra random params
- [x] Automatically open map generator in scenario editor
- [x] Make random simplex noise map the default generator? -> reverted
- [x] Prevent snow terrain edges underneath non-snow terrain
- [ ] ~~Delete old `map_gen_noise.png` and `map_gen_trees.png`? (NB: exposed through `scripting/IconNames.hpp` -- but are they used by any scripts?)~~
- [ ] Add optional automatic assignment of park land? -> later
- [ ] Add optional automatic placement of park entrance, path, and guest spawn points? -> later

J-PIE-314 commented:
Yes! Here I am on this page thinking about how the window would look better with the trees icon from the scenery selection window, and "bam!", you updated it. Good work. 👍

tupaschoal commented:
On the first tab
- If you select "Heightmap source" as "Heightmap file" the generate button is deactivated. It does not get enabled again after returning to "flatland" or "simplex noise", only if you switch to any of the other 3 tabs and come back.

On the second tab:
- seems like the checkbox for "random terrain" spans the whole line, and it's really weird to change the terrain type
https://github.com/user-attachments/assets/6bf7cc26-64b2-4f3c-becd-27105663b37f

On the third tab:
- Sometimes it seems like it generates beaches around no water:
![image](https://github.com/user-attachments/assets/5bd26f8f-5fb3-4b58-bf7b-6daa264d7ab7)

On the fourth tab:
- I found it weird that tree to land ratio cannot be lower than 0%.
- If you go down all the way to max tree altitude as -9, the min tree one gets to -10
![image](https://github.com/user-attachments/assets/cae17920-c31e-419c-b262-1dfd7289e028)

AaronVanGeffen commented:
Thanks for testing, @tupaschoal!

> If you select "Heightmap source" as "Heightmap file" the generate button is deactivated. It does not get enabled again after returning to "flatland" or "simplex noise", only if you switch to any of the other 3 tabs and come back.

Nice catch. Fixed!

> seems like the checkbox for "random terrain" spans the whole line, and it's really weird to change the terrain type

So glad you caught this! I had noticed it before, but forgot all about it. In fact, the widgets still had a temporary offset in their Y definitions, suggesting I forgot to finish them. Anyway, fixed now.

> Sometimes it seems like it generates beaches around no water:

Not sure what's causing that, tbh... This PR only aims to improve the UI, though. Could you file an issue for this?

> I found it weird that tree to land ratio cannot be lower than 0%.

I'm not sure what a negative ratio would mean, in this case. To me, 0% means no tree coverage, where 100% means the land is pretty much entirely covered in trees. Does that make sense?

> If you go down all the way to max tree altitude as -9, the min tree one gets to -10

Nice catch! This was an off-by-one error. Should be fixed now.

tupaschoal commented:
> I'm not sure what a negative ratio would mean, in this case. To me, 0% means no tree coverage, where 100% means the land is pretty much entirely covered in trees. Does that make sense?

I mistyped, I think upon testing I could never get it below 10%

AaronVanGeffen commented:
> I mistyped, I think upon testing I could never get it below 10%

Ah, I see! When I introduced the parameter, it looks like I translated the old `treeToLandRatio = (10 + (UtilRand() % 30)) / 100.0f;` to a range of 10–40% and then increased it to 50% max. I see no reason to go below that, necessarily, though 0% would effectively turn the option off entirely.

Maybe 1% would be better as a minimum, rather than 10% or indeed 0%?

tupaschoal commented:
I think we should allow the full range, 0-100%, although I understand 100% will look stupid.

In terms of UI, maybe if user goes to 0 we uncheck the check box? Or we drop the checkbox entirely and make it controllable only via the percentage.

AaronVanGeffen commented:
> In terms of UI, maybe if user goes to 0 we uncheck the check box? Or we drop the checkbox entirely and make it controllable only via the percentage.

I think neither of those would make for good UI, to be honest. For that reason, I prefer sticking with the 1% minimum.

> I think we should allow the full range, 0-100%, although I understand 100% will look stupid.

Initially, I was thinking I would be fine allowing an increase to 100%, but the way the placement is set up, it's always placing around noise:

![Unnamed park 2024-09-19 23-05-21](https://github.com/user-attachments/assets/595a1bc0-1c41-427e-844c-a1ba21a6d339)

I'm not sure this would work intuitively if we do provide the full range. Perhaps we can improve this further in the future. For now, I would prefer not to touch this part of the code.

tupaschoal commented:
Fair enough

tupaschoal commented:
Thinking more about the UI I do think sliders would be better than spin boxes for most of the options, but in fine if you don't wanna change it either

AaronVanGeffen commented:
> Thinking more about the UI I do think sliders would be better than spin boxes for most of the options, but in fine if you don't wanna change it either

Yeah, I agree that would provide more intuitive controls. In OpenRCT2, we don't really have sliders, though, just scroll bars that have been tricked into believing they are for the purpose of volume control. I'd like us to port actual sliders one day, like in Locomotion/OpenLoco, but that's definitely not for this PR.

Gymnasiast reviewed (changes requested):
I think this is a tremendous improvement, cutting through layer of paint and filler to make the window much easier to understand.

Going through the interface with a fine comb, I did hit upon some remaining snags.

1. This dropdown is not wide enough and could use the "Stay open" flag.
![afbeelding](https://github.com/user-attachments/assets/85b56325-31a3-42f3-8790-e2797275ac1b)

2. When unchecking "Place trees", the options below should be greyed out.
![afbeelding](https://github.com/user-attachments/assets/40e2d823-3490-4dc9-b102-0c696298c44b)

3. Height steppers does not show half-metres correctly (jumps from 6 to 7 to 9, rather than 6 to 7.5 to 9)
![afbeelding](https://github.com/user-attachments/assets/4ec85840-0f43-49af-bccd-75fbcd435a17)

4. Change “height source” immediately regenerates terrain, causing a slowdown. I think it should only regenerate when the user clicks that button. It’s more predictable that way.
![afbeelding](https://github.com/user-attachments/assets/eb987443-d735-425a-a379-93e275e5c972)

5. I can’t for the life of me figure out why anyone would want to disable this option:
![afbeelding](https://github.com/user-attachments/assets/c4d16947-e566-4366-9443-03f411789151)

6. These number inputs do not allow clicking:
![afbeelding](https://github.com/user-attachments/assets/c689605d-e49b-4ff8-b3e6-91b5d951ba8f)
![afbeelding](https://github.com/user-attachments/assets/7f6a5cc3-1730-437a-8080-bbca7ec6d53c)

7. Some options do nothing on flatland, but are not hidden:
![afbeelding](https://github.com/user-attachments/assets/7d314a32-6499-4b80-9be5-65336f78ca29)
![afbeelding](https://github.com/user-attachments/assets/5875bcf4-f5da-454f-9e5c-a723476b5d3e)
![afbeelding](https://github.com/user-attachments/assets/a89cfe8a-b3f5-4470-8042-27e70b51b18c)
![afbeelding](https://github.com/user-attachments/assets/20d10da1-8c6f-48bc-a32b-ed0c55f576fb)

AaronVanGeffen commented:
Thanks for the thorough review.

> This dropdown is not wide enough and could use the "Stay open" flag.

Interestingly, it used to use the full width. I think I messed it up in a rebase. Will fix. Applying that flag sounds like a good idea.

> When unchecking "Place trees", the options below should be greyed out.

> Some options do nothing on flatland, but are not hidden:

Agreed, will take care of that.

> Height steppers does not show half-metres correctly (jumps from 6 to 7 to 9, rather than 6 to 7.5 to 9)

You're right, and that indeed has to do with rounding. I think I should switch from UINT16 to COMMA16_1DP formatting codes, but I was having trouble getting it to work previously. Worth fixing, though, indeed.

> Change “height source” immediately regenerates terrain, causing a slowdown. I think it should only regenerate when the user clicks that button. It’s more predictable that way.

Personally, I like that it immediately regenerates, particularly during testing. I think you're right that it might be better not to, though. The third option can't regenerate without an image, anyway, so that would make it more consistent.

> I can’t for the life of me figure out why anyone would want to disable [the smooth tile edges] option

For terrain, I can't imagine doing so either, no. I imagine it's there for images that require sharp edges, like with letters or other novelty maps. Best to leave it, lest it break someone's workflow, I thought.

> These number inputs do not allow clicking:

Yeah, the way the text input 'on click' works is rather cumbersome to implement. I've already added it in other places, though, so I suppose I might as well tackle these as well.

AaronVanGeffen commented:
@Gymnasiast I think that should be your concerns addressed. As discussed on Discord, let's leave the `{LENGTH}` format business for a future PR.

Broxzier reviewed (approved):
Gave this a spin and I found that the updated interface behaves exactly like expected, and the new icons fit well too. Great job!

The only nitpicky thing I ran into is that when selecting a heightmap as source, the map size spinners can still be used, but they are entirely ignored. If it's not too much work, I'd like to see them disable or even made invisible, as their value also doesn't reflect the map size anymore.

AaronVanGeffen commented:
> Gave this a spin and I found that the updated interface behaves exactly like expected, and the new icons fit well too. Great job!

Thank you! That's great to hear.

> The only nitpicky thing I ran into is that when selecting a heightmap as source, the map size spinners can still be used, but they are entirely ignored. If it's not too much work, I'd like to see them disable or even made invisible, as their value also doesn't reflect the map size anymore.

That's a good point. You're right, that's confusing UI. I'll make the change.

Gymnasiast commented:
> For terrain, I can't imagine doing so either, no. I imagine it's there for images that require sharp edges, like with letters or other novelty maps. Best to leave it, lest it break someone's workflow, I thought.

![relevant XKCD](https://imgs.xkcd.com/comics/workflow.png)

AaronVanGeffen commented:
> > The only nitpicky thing I ran into is that when selecting a heightmap as source, the map size spinners can still be used, but they are entirely ignored. If it's not too much work, I'd like to see them disable or even made invisible, as their value also doesn't reflect the map size anymore.
> 
> That's a good point. You're right, that's confusing UI. I'll make the change.

@Broxzier Made the changes. The window now looks like this until a height map image is selected:

<img width="471" alt="Screenshot 2024-09-22 at 21 29 04" src="https://github.com/user-attachments/assets/957ad8a6-8d81-4bd3-b3d7-ce488244b45f">

Gymnasiast reviewed (changes requested):
1.
> You're right about some of these, but min/max land height does affect flatland (only min though), as does water level and min/max tree altitude (again, only min). I prefer to leave these as-is.

There should be only one input then, not two, and it should have a modified label.

2.  You did not address the “Smooth tile edges” option. 
- On flatland, it does nothing
- On Simplex, you’d have to be high on crack to turn it off
- On heightmaps, it _may_ be useful, although we already have Normalise and Smooth Height Map, which do related things and have very similar naming.

![afbeelding](https://github.com/user-attachments/assets/5c55512f-3279-4645-bd40-0a3fa6c6c31f)

3. The heightmap options immediately take effect, rather than waiting for the user to click Generate. I think this should be taken out too, both for consistency and also to avoid slowing down the user when they want to tweak a few parameters in a row.

4. Normalise Height Map is right next to the button to select a heightmap, suggesting that this option controls where you can select one or not, which is not the case Rather, they should be above each other.

5. There is no indication as to how “Normalise height map” and “Smooth height map” relate. This is especially unclear considering their similar naming. There are no tooltips either.

6. The strength input of the heightmap does not allow input by click.

Sorry for the points about heightmaps, I hadn’t tested those earlier.

AaronVanGeffen commented:
> > You're right about some of these, but min/max land height does affect flatland (only min though), as does water level and min/max tree altitude (again, only min). I prefer to leave these as-is.
> 
> There should be only one input then, not two, and it should have a modified label.

Not sure I agree. I thought about keeping the 'Base height' label around for this purpose, but I don't think it makes it clearer. Personally, I prefer to have a predictable interface, with options staying in the same place.

> 2. You did not address the “Smooth tile edges” option.
> 
> * On flatland, it does nothing
> * On Simplex, you’d have to be high on crack to turn it off
> * On heightmaps, it _may_ be useful, although we already have Normalise and Smooth Height Map, which do related things and have very similar naming.

What do you suggest we do with it?

> 3. The heightmap options immediately take effect, rather than waiting for the user to click Generate. I think this should be taken out too, both for consistency and also to avoid slowing down the user when they want to tweak a few parameters in a row.

Admittedly, I hadn't considered that. I'll look into changing it.

> 4. Normalise Height Map is right next to the button to select a heightmap, suggesting that this option controls where you can select one or not, which is not the case Rather, they should be above each other.

That's a good point. I think I'll change it to be more like OpenLoco's, which shows the currently selected filename:

![openloco_landscape_gen_0b](https://github.com/user-attachments/assets/cd135d6b-f48d-4ac8-950f-540ebf6280af)

> 5. There is no indication as to how “Normalise height map” and “Smooth height map” relate. This is especially unclear considering their similar naming. There are no tooltips either.

I don't disagree, but please realise I didn't introduce these options. This feels like moving the goal posts a bit much. I agree tooltips would be better, but other fields could use them as well. For a follow-up PR, imo.

> 6. The strength input of the heightmap does not allow input by click.

There just had to be one more... 🥲 I'll address it.

AaronVanGeffen commented:
@Gymnasiast The height map settings pane now shows the currently selected height map filename, with a 'Browse...' button next to it:

![Unnamed park 2024-09-23 14-49-30](https://github.com/user-attachments/assets/ebd8e0bd-4f39-4a2f-947e-47ae8aecc0ba)

AaronVanGeffen commented on src/openrct2-ui/windows/MapGen.cpp:
```diff
@@ -1351,41 +1474,74 @@ namespace OpenRCT2::Ui::Windows
-                SetPage(WINDOW_MAPGEN_PAGE_HEIGHTMAP);
-
-                HeightmapGenerateMap();
+                _heightmapFilename = std::string(fs::u8path(path).filename());
```
Our Windows CI is complaining about this line. I added the explicit `std::string` constructor call in an attempt to mitigate this, but clearly to no avail. Any ideas?

tupaschoal replied:
```suggestion
                _heightmapFilename = fs::u8path(path).filename().string();
```
Should do the trick

Gymnasiast reviewed (changes requested):
1. The map is still autogenerated when changing a heightmap option.
2. I still think the height options should be modified for flatland, what is currently here plainly does not make sense. A consistent user interface means nothing if it confuses people.
3. The "smooth edges" option should be hidden for flatland and simplex, rather than just disabled.

AaronVanGeffen commented:
> 1. The map is still autogenerated when changing a heightmap option.

Yes, I mentioned it in my rebuttal above. It was intentional on my part.

Thinking about it, I can remove the autogenerating now, then re-introduce it in a future PR along with a progress bar, perhaps?

> 2. I still think the height options should be modified for flatland, what is currently here plainly does not make sense. A consistent user interface means nothing if it confuses people.

Okay, we need a compromise here then. I would rather not add different text labels, because that means three extra strings per field (including text input). As a compromise, I can just disable the 'max' fields, leaving the 'min' fields enabled?

> 3. The "smooth edges" option should be hidden for flatland and simplex, rather than just disabled. 

I would really rather not, tbh. Disabling them suggests that these options exist at all, which hiding does not.

AaronVanGeffen commented:
@Gymnasiast I have removed the lines that triggered regenerating the map automatically for the heightmap settings, and disabled the 'max' terrain and forest height widgets in 'flatland' mode:

<img width="462" alt="Screenshot 2024-09-29 at 20 01 42" src="https://github.com/user-attachments/assets/c2be454b-0abe-43be-800b-63c56386ae7e">
<img width="462" alt="Screenshot 2024-09-29 at 20 01 47" src="https://github.com/user-attachments/assets/d7f335e2-fa51-47e8-a128-7f422ed837f1">

Gymnasiast reviewed (approved):
Not completely happy, but this has been dragging for too long. It’s still a noticeable improvement.

AaronVanGeffen commented:
Okay, thanks. I'll try to get this rebased tonight then.

Gymnasiast commented:
I feel I should clarify that this essentially achieves 90% of what I wanted, just not quite as "perfect" as I wanted it to be. It is still a massive improvement, and I also feel like I have to be careful not to push you beyond breaking point. Nobody wins if it burns you out.
